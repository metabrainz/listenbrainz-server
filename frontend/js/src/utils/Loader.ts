import type { LoaderFunctionArgs, Params } from "react-router-dom";
import { json } from "react-router-dom";
import { isEmpty } from "lodash";
import queryClient from "./QueryClient";
import { getObjectForURLSearchParams } from "./utils";

const RouteLoader = async ({ request }: { request: Request }) => {
  const response = await fetch(request.url, {
    method: "POST",
    headers: {
      Accept: "application/json",
    },
  });
  const data = await response.json();
  if (!response.ok) {
    throw json(data, { status: response.status });
  }
  return data;
};

export default RouteLoader;

const RouteLoaderURL = async (url: string) => {
  const response = await fetch(url, {
    method: "POST",
    headers: {
      Accept: "application/json",
    },
  });
  const data = await response.json();
  if (!response.ok) {
    throw json(data, { status: response.status });
  }
  return data;
};

export const RouteQuery = (key: any[], url: string) => ({
  queryKey: key,
  queryFn: async () => {
    const data = await RouteLoaderURL(url);
    return data;
  },
});

export const RouteQueryLoader = (
  routeKey: string,
  includeSearchParams = false
) => async ({ request, params }: LoaderFunctionArgs) => {
  const keys = [routeKey] as any[];

  // Add params to the keys
  const paramsObject = { ...params };
  if (!isEmpty(paramsObject)) keys.push(paramsObject);

  if (includeSearchParams) {
    // Add search params to the keys
    const searchParams = new URLSearchParams(request.url.split("?")[1]);
    const searchParamsObject = getObjectForURLSearchParams(searchParams);
    keys.push(searchParamsObject);
  }

  await queryClient.ensureQueryData(RouteQuery(keys, request.url || ""));
  return null;
};
