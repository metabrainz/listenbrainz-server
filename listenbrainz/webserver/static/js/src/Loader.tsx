import * as React from "react";
import { faSpinner } from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { IconProp } from "@fortawesome/fontawesome-svg-core";

type LoaderProps = {
  isLoading: boolean;
};

const Loader: React.SFC<LoaderProps> = (props) => {
  // eslint-disable-next-line react/prop-types
  const { isLoading, children } = props;
  return isLoading ? (
    <FontAwesomeIcon icon={faSpinner as IconProp} size="4x" spin />
  ) : (
    <>{children}</>
  );
};

export default Loader;
