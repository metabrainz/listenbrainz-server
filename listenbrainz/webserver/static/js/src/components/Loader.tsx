import * as React from "react";
import Spinner from "react-loader-spinner";

type LoaderProps = {
  isLoading: boolean;
  style?: React.CSSProperties;
  className?: string;
  loaderText?: string;
  [key: string]: any;
};

export default function Loader(props: React.PropsWithChildren<LoaderProps>) {
  const {
    isLoading,
    children,
    className,
    loaderText,
    style: propStyle,
    ...rest
  } = props;

  const style: React.CSSProperties = {
    display: "flex",
    flexDirection: "column",
    alignItems: "center",
    justifyContent: "center",
    minHeight: "inherit",
    ...propStyle,
  };

  return isLoading ? (
    <div className={`text-center ${className || ""}`} style={style} {...rest}>
      <Spinner type="Oval" color="#cccccc" height={30} width={30} />
      {loaderText && (
        <small>
          <p className="text-muted mt-5">{loaderText}</p>
        </small>
      )}
    </div>
  ) : (
    <>{children}</>
  );
}
