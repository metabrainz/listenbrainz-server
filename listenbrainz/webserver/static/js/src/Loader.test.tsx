import * as React from "react";
import { mount } from "enzyme";

import Loader from "./Loader";

const ChildComponent = () => {
  return <div>Child Component</div>;
};

describe("Loader", () => {
  it('renders loader when "isLoading" is true', () => {
    const wrapper = mount(
      <Loader isLoading>
        <ChildComponent />
      </Loader>
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('renders child component when "isLoading" is false', () => {
    const wrapper = mount(
      <Loader isLoading={false}>
        <ChildComponent />
      </Loader>
    );

    expect(wrapper).toMatchSnapshot();
  });
});
