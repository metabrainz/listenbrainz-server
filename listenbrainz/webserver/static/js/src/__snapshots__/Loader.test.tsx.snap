// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loader renders child component when "isLoading" is false 1`] = `
<Loader
  isLoading={true}
>
  <FontAwesomeIcon
    border={false}
    className=""
    fixedWidth={false}
    flip={null}
    icon={
      Object {
        "icon": Array [
          512,
          512,
          Array [],
          "f110",
          "M304 48c0 26.51-21.49 48-48 48s-48-21.49-48-48 21.49-48 48-48 48 21.49 48 48zm-48 368c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm208-208c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zM96 256c0-26.51-21.49-48-48-48S0 229.49 0 256s21.49 48 48 48 48-21.49 48-48zm12.922 99.078c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.491-48-48-48zm294.156 0c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.49-48-48-48zM108.922 60.922c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.491-48-48-48z",
        ],
        "iconName": "spinner",
        "prefix": "fas",
      }
    }
    inverse={false}
    listItem={false}
    mask={null}
    pull={null}
    pulse={false}
    rotation={null}
    size="4x"
    spin={true}
    swapOpacity={false}
    symbol={false}
    title=""
    transform={null}
  >
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-spinner fa-w-16 fa-spin fa-4x "
      data-icon="spinner"
      data-prefix="fas"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 512 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M304 48c0 26.51-21.49 48-48 48s-48-21.49-48-48 21.49-48 48-48 48 21.49 48 48zm-48 368c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm208-208c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zM96 256c0-26.51-21.49-48-48-48S0 229.49 0 256s21.49 48 48 48 48-21.49 48-48zm12.922 99.078c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.491-48-48-48zm294.156 0c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.49-48-48-48zM108.922 60.922c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.491-48-48-48z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
  </FontAwesomeIcon>
</Loader>
`;

exports[`Loader renders loader when "isLoading" is true 1`] = `
<Loader
  isLoading={true}
>
  <FontAwesomeIcon
    border={false}
    className=""
    fixedWidth={false}
    flip={null}
    icon={
      Object {
        "icon": Array [
          512,
          512,
          Array [],
          "f110",
          "M304 48c0 26.51-21.49 48-48 48s-48-21.49-48-48 21.49-48 48-48 48 21.49 48 48zm-48 368c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm208-208c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zM96 256c0-26.51-21.49-48-48-48S0 229.49 0 256s21.49 48 48 48 48-21.49 48-48zm12.922 99.078c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.491-48-48-48zm294.156 0c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.49-48-48-48zM108.922 60.922c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.491-48-48-48z",
        ],
        "iconName": "spinner",
        "prefix": "fas",
      }
    }
    inverse={false}
    listItem={false}
    mask={null}
    pull={null}
    pulse={false}
    rotation={null}
    size="4x"
    spin={true}
    swapOpacity={false}
    symbol={false}
    title=""
    transform={null}
  >
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-spinner fa-w-16 fa-spin fa-4x "
      data-icon="spinner"
      data-prefix="fas"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 512 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M304 48c0 26.51-21.49 48-48 48s-48-21.49-48-48 21.49-48 48-48 48 21.49 48 48zm-48 368c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm208-208c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zM96 256c0-26.51-21.49-48-48-48S0 229.49 0 256s21.49 48 48 48 48-21.49 48-48zm12.922 99.078c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.491-48-48-48zm294.156 0c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.49-48-48-48zM108.922 60.922c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.491-48-48-48z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
  </FontAwesomeIcon>
</Loader>
`;
